docs/dev/replay-e2e-test.md

```diff
@@ -0,0 +1,51 @@
+
+These tests should work properly without any real connection settings.
+
+## 3 Test Modes
+
+There are 3 representative values of the environment variable `PROMPT_FLOW_TEST_MODE`
+- `live`: Tests run against the real backend, which is the way traditional end-to-end tests do.
+- `record`: Tests run against the real backend, and network traffic will be sanitized (filter sensitive and unnecessary requests/responses) and recorded to local files (recordings).
+- `replay`: There is no real network traffic between SDK/CLI and the backend, tests run against local recordings.
+
+## How to Record a Test
+
+To record a test, don’t forget to clone the full repo and set up the proper test environment following [dev_setup.md](./dev_setup.md):
+1. Prepare some data.
+   * If you have changed/affected tests in __sdk_cli_test__: Copy or rename the file [dev-connections.json.example](../../src/promptflow/dev-connections.json.example) to `connections.json` in the same folder.
+   * If you have changed/affected tests in __sdk_cli_azure_test__: Prepare a config.json, point your `az``  to the proper tenant, subscription, resource group, and workspace.
+2. Record the test.
+   * Specify the environment variable `PROMPT_FLOW_TEST_MODE` to `'record'`. If you have a `.env` file, we recommend specifying it there. Then, just run the test that you want to record.
+3. Once the test completed.
+   * If you have changed/affected tests in __sdk_cli_azure_test__: There should be one new YAML file located in `src/promptflow/tests/test_configs/recordings/`, containing the network traffic of the test.
+   * If you have changed/affected tests in __sdk_cli_test__: There may be changes in the folder `src/promptflow/tests/test_configs/node_recordings/`.  Don’t worry if there are no changes, because similar LLM calls may have been recorded before.
+
+## Techniques Behind Replay Test
+
+### sdk_cli_azure_test
+
+End-to-end tests for pfazure aim to test the behavior of the PromptFlow SDK/CLI as it interacts with the service. This process can be time-consuming, error-prone, and require credentials (which are unavailable to pull requests from forked repositories); all of these go against our intention for a smooth development experience.
+
+Therefore, we introduce replay tests, which leverage [VCR.py](https://pypi.org/project/vcrpy/) to record all required network traffic to local files and replay during tests. In this way, we avoid the need for credentials, speed up, and stabilize the test process.
+
+### sdk_cli_test
+
+sdk_cli_test often doesn’t use a real backend. It will directly invoke LLM calls from localhost. Thus the key target of replay tests is to avoid the need for OpenAI tokens. If you have OpenAI / Azure OpenAI tokens yourself, you can try recording the tests. Record Storage will not record your own LLM connection, but only the input and output of the LLM calls.
```

Comment:
```suggestion
sdk_cli_test often doesn’t use a real backend. It will directly invokes LLM calls from localhost. Thus the key target of replay tests is to avoid the need for OpenAI tokens. If you have OpenAI / Azure OpenAI tokens yourself, you can try recording the tests. Record Storage will not record your own LLM connection, but only the input and output of the LLM calls.
```